src/gt4sd/frameworks/gflownet/train/core.py

```diff
@@ -0,0 +1,147 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+"""Train module implementation."""
+
+import logging
+from argparse import Namespace
+from typing import Any, Dict
+
+import sentencepiece as _sentencepiece
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.loggers import TensorBoardLogger
+
+from ..arg_parser.parser import parse_arguments_from_config
+
+# from ..dataloader import build_dataset
+from ..dataloader.data_module import GFlowNetDataModule
+from ..dataloader.dataset import GFlowNetDataset, GFlowNetTask
+
+# from ..envs import build_env_context
+from ..envs.graph_building_env import GraphBuildingEnv, GraphBuildingEnvContext
+from ..loss import ALGORITHM_FACTORY
+from ..ml.models import MODEL_FACTORY
+from ..ml.module import GFlowNetModule
+
+# from ..train import build_task
+
+# sentencepiece has to be loaded before lightning to avoid segfaults
+_sentencepiece
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+
+def train_gflownet(
+    configuration: Dict[str, Any],
+    dataset: GFlowNetDataset,
+    environment: GraphBuildingEnv,
+    context: GraphBuildingEnvContext,
+    task: GFlowNetTask,
+) -> None:
+    """Train a gflownet given a configuration, a dataset and a task.
+    The default enviroment and context are compatible with small molecules.
+
+    Args:
+        configuration: a configuration dictionary.
+        dataset: a dataset compatible with lightning.
+        environment: an environment specifying the state space.
+        context: an environment context specifying how to combine states.
+        task: a task specifying the reward structure.
+    """
+
+    arguments = Namespace(**configuration)
+
+    algorithm = ALGORITHM_FACTORY[getattr(arguments, "algorithm")](
+        configuration=configuration,
+        environment=environment,
+        context=context,
+    )
+    model = MODEL_FACTORY[getattr(arguments, "model")](
+        configuration=configuration,
+        context=context,
+    )
+
+    dm = GFlowNetDataModule(
+        configuration=configuration,
+        dataset=dataset,
+        environment=environment,
+        context=context,
+        task=task,
+        algorithm=algorithm,
+        model=model,
+    )
+    dm.prepare_data()
+
+    module = GFlowNetModule(
+        configuration=configuration,
+        dataset=dataset,
+        environment=environment,
+        context=context,
+        task=task,
+        algorithm=algorithm,
+        model=model,
+    )
+
+    tensorboard_logger = TensorBoardLogger(
+        "logs", name=getattr(arguments, "basename", "default")
+    )
+    checkpoint_callback = ModelCheckpoint(
```

Comment:
Do you want to save the weights for the last epoch only. Or to monitor some metric/loss and save also the best one?

Reply:
I want to monitor and save the best one, but I don't know how expensive/time-consuming is to run validation every x-epochs (I only run short training for debugging until now).

Reply:
You can start with monitoring the training loss until you'll have a validation loss/metric.